### ISSUE TITLE
How to get the next item from graphql?

### ISSUE THREAD

**Issue by SteveEdson** [gridsome/gridsome#177]
I get get all posts on my site with the following query, and have it show the next item in the array.

```graphql
query {
  allPost {
    edges {
      node {
        title
      }
      next {
        title
      }
    }
  }
}
```

However, when on an individual post, with the following query:

```graphql
query Post {
    post (path: "/path") {
        title
    	content        
    }
}
```

How can I get the next and previous items? Is there a way to get the specified post using the `allPost` query and regex? Any help would be appreciated.

Thanks

**Comment by hjvedvik** [comment]
Currently, there is no easy way to do that, unfortunately. But we can consider implementing `nextNode` and `previousNode` fields on each node type. And also have an argument to decide if the next node should be ordered by something other than the date.

**Comment by SteveEdson** [comment]
Perfect, thanks!

**Comment by kaceo** [comment]
Yes please this would make it much easier to convert a pile of pages into a manual, simplifying the process of reading from first to last chapter in sequence.  

I notice that even Gridsome's own website documentation has to use a separate data yaml file to define the table of content because Graphql does not thread all the pages together.

**Comment by hjvedvik** [comment]
Would this be a good solution? Default sorting is by descending `date`, so `next` will be older posts. But the field can have custom sorting with `sortBy` and `order` arguments like `edges (sortBy: "title", order: ASC) { ... }`.

```graphql
query Post ($id: String) {
  myPost (id: $id) {
    title
    edges {
      previous {
        title
      }
      next {
        title
      }
    }
  }
}
```

**Comment by kaceo** [comment]
Date sorting probably makes sense for blogs and news but for software documentation pages the previous and next should be sorted based on section-chapter-pages hierarchy.  I would suggest the source be organised by directory file name sort order (alphabetic) so we can use, say, 100-Introduction and 200-How-to-Install to change the sequence of pages/folders.  It will also make it easier to insert pages without having to massively edit lots of files.  The front matter of each file can contain the actual correct title.

While we are on the subject, I like to also address the multi-lingual file issue, as this is already tackled by Vuepress.  They allow a content file to have sub-directories with different languages.   We may want to at least be thinking ahead of the potential issues affecting Slugs and Paths.

**Comment by hjvedvik** [comment]
Sort by directory structure order might not work for source nodes from external APIs. Unless the filesystem source plugin inserted and resolved this field itself. But what about implementing a more advanced `sort` argument to sort on multiple fields? This example would sort by a `group` field first, then by a `weight` field:

```graphql
query Post ($id: String) {
  myPost (id: $id) {
    title
    edges (sort: [{ by: "group", order: ASC }, { by: "weight", order: ASC  }]) {
      previous {
        title
      }
      next {
        title
      }
    }
  }
}
```

Default sort would be `sort: [{ by: "date", order: DESC }]`.

**Comment by hellocosmin** [comment]
Really looking forward to this - working on some blog starters where it would be very useful. Can't wait! 😀

**Comment by hjvedvik** [comment]
I think we'll add the `edges.next/previous` example above as a simple solution. But the `@gridsome/source-filesystem` will get an overhaul sometime soon to support the number prefixes in filenames and more advanced next previous handling based on file structure.

**Comment by hellocosmin** [comment]
Awesome! Working on a blog starter for Gridsome as we speak, will be great to do next/prev collection item queries right in a post template 👍

**Comment by robaxelsen** [comment]
@hjvedvik any progress on this?

**Comment by hjvedvik** [comment]
I have been testing various options and after some consideration, I think we will wait with this. We are going to allow more configuration for how pages are created from nodes in Gridsome 0.7 and that will probably be a better way to enable next and prev navigation.

As a workaround in 0.6, you can use the new API for creating pages and set the ids for next and prev node as query variables. Just rename the template file to something else than the type name to prevent auto-created pages. Here is an example:

```js
module.exports = function (api) {
  api.createPages(({ graphql, createPage }) => {
    const result = await graphql('query sorted posts')

    result.data.allPost.edges.forEach(({ node }, i, edges) => {
      const prev = edges[i - 1]
      const next = edges[i + 1]
    
      createPage({
        path: edge.node.path,
        component: './src/templates/PostTemplate.vue'
        queryVariables: {
          id: edge.node.id,
          prevtId: prev ? prev.node.id : null,
          nextId: next ? next.node.id : null
        }
      })
    })
  })
}
```
And in the `page-query` in `./src/templates/PostTemplate.vue`:
```graphql
query Post ($id: String, $prevId: String, $nextId: String) {
  post (id: $id) { ... }
  prev: post (id: $prevId) { ... }
  next: post (id: $nextId) { ... }
}
```

**Comment by kayakmind** [comment]
Not sure where this stands, but I'm a vote in favor!

**Comment by derron1** [comment]
> Not sure where this stands, but I'm a vote in favor!

it currently beening prioritized for version 0.70

**Comment by SteveEdson** [comment]
Sorry to bump this again, has this been delayed? Still really keen to be able to use this

**Comment by hacknug** [comment]
@SteveEdson I have a plugin (haven't published yet) that works around this like @hjvedvik suggested. Then only issue with it is i get a `Error: Duplicate key for property path` when I don't set a different one than `edge.node.path`. Since the plugin allows you to extend/configure the `createPage` options, it's not a deal breaker (although things don't look as clean as they should/could).

Let me know if you'd want me to publish it 👍

**Comment by rubik** [comment]
Any updates on this? I see the milestone was deleted, so it's not clear if this will make it in the next release or not.

**Comment by adalmagro** [comment]
Hi, are there any updates on this? I am using Gridsome + Contentful as CMS. Anybody can tell how it is implemented within the Gridsome own docs
<img width="760" alt="Captura de pantalla 2022-03-17 a las 12 18 51" src="https://user-images.githubusercontent.com/504739/158798259-93a8d2f3-338e-4100-ba95-3e03c97bdfd2.png">
?
